transaction/sort.go

```diff
@@ -0,0 +1,78 @@
+package transaction
+
+import (
+	"container/heap"
+	"sort"
+)
+
+var DefaultCompare = CompareTxnsByFeePerSize
+
+func DefaultSort(s []*Transaction) sort.Interface {
+	return SortTxnsByFeePerSize(s)
+}
+
+func DefaultHeap(s []*Transaction) heap.Interface {
+	return (*SortTxnsByFeePerSize)(&s)
+}
+
+type SortTxnsByNonce []*Transaction
+
+func (s SortTxnsByNonce) Len() int           { return len(s) }
+func (s SortTxnsByNonce) Swap(i, j int)      { s[i], s[j] = s[j], s[i] }
+func (s SortTxnsByNonce) Less(i, j int) bool { return s[i].UnsignedTx.Nonce < s[j].UnsignedTx.Nonce }
+
+func CompareTxnsByFee(txn1, txn2 *Transaction) int {
+	if txn1.UnsignedTx.Fee > txn2.UnsignedTx.Fee {
+		return 1
+	}
+	if txn1.UnsignedTx.Fee < txn2.UnsignedTx.Fee {
+		return -1
+	}
+	if txn1.GetSize() > txn2.GetSize() {
+		return -1
+	}
+	if txn1.GetSize() < txn2.GetSize() {
+		return 1
+	}
+	return 0
+}
+
+type SortTxnsByFee []*Transaction
+
+func (s SortTxnsByFee) Len() int            { return len(s) }
+func (s SortTxnsByFee) Swap(i, j int)       { s[i], s[j] = s[j], s[i] }
+func (s SortTxnsByFee) Less(i, j int) bool  { return CompareTxnsByFee(s[i], s[j]) < 0 }
+func (s *SortTxnsByFee) Push(x interface{}) { *s = append(*s, x.(*Transaction)) }
+func (s *SortTxnsByFee) Pop() interface{} {
+	old := *s
+	n := len(old)
+	x := old[n-1]
+	*s = old[0 : n-1]
+	return x
+}
+
+func CompareTxnsByFeePerSize(txn1, txn2 *Transaction) int {
+	feePerSize1 := float64(txn1.UnsignedTx.Fee) / (float64(txn1.GetSize()) + 0.01)
```

Comment:
What does this `0.01` do?

Reply:
Prevent panic when txn size is zero, it shouldn't happen, just prevent malicious transaction

Reply:
> 

But malicious transaction with zero size will have higher rank with very small tx fee

Reply:
No valid transaction will have be smaller than generate ID transaction v1. This is just to prevent invalid transaction from making node panic